File: Model/MatchGenerators/MatchDaysMatchGenerator.vb

Public Class MatchdaysMatchGenerator
    Implements IMatchGenerator
    Private ReadOnly _teamsCount As Integer
    Private ReadOnly _matches As New List(Of MatchIndexPair)()
    Private ReadOnly _matchdays As New List(Of Matchday)()

    Public Sub New(Optional teamsCount As Integer = 10)
        _teamsCount = teamsCount
    End Sub

    Public Function Generate() As IEnumerable(Of MatchIndexPair) Implements IMatchGenerator.Generate
        _matches.Clear()
        _matches.AddRange(GenerateMatches())
        _matchdays.Clear()
        _matchdays.AddRange(GenerateMatchdays())
        Return MatchIndexPairsOrderdByMatchdays()
    End Function

    Private Iterator Function MatchIndexPairsOrderdByMatchdays() As IEnumerable(Of MatchIndexPair)
        For Each matchday In _matchdays
            For Each matchIndexPair In matchday
                Yield matchIndexPair
            Next matchIndexPair
        Next matchday
    End Function

    Private Iterator Function GenerateMatches() As IEnumerable(Of MatchIndexPair)
        For i As Integer = 1 To _teamsCount
            For j As Integer = i + 1 To _teamsCount
                Yield New MatchIndexPair(i, j)
            Next
        Next
    End Function

    Private Iterator Function GenerateMatchdays() As IEnumerable(Of Matchday)
        Dim blockSize As Integer = _teamsCount \ 2

        For i As Integer = 0 To _teamsCount - 2
            Yield GenerateMatchday(blockSize)
        Next
    End Function

    Private Function GenerateMatchday(blockSize As Integer) As Matchday
        Dim matchday As New Matchday()
        While matchday.Count < blockSize
            Dim invalidatedMatches As New List(Of MatchIndexPair)()
            Dim nextMatch As MatchIndexPair = GetNextMatch(matchday)
            If nextMatch.Equals(MatchIndexPair.Empty) Then
                nextMatch = _matches.Except(invalidatedMatches).FirstOrDefault()
                If nextMatch Is Nothing Then nextMatch = MatchIndexPair.Empty
                invalidatedMatches.AddRange(matchday.MatchesWithPlayersFrom(nextMatch))
                _matches.AddRange(invalidatedMatches)
                matchday.RemoveAll(Function(m) invalidatedMatches.Contains(m))
            End If
            matchday.Add(nextMatch)
            _matches.Remove(nextMatch)
        End While
        Return matchday
    End Function

    Private Function GetNextMatch(matchday As Matchday) As MatchIndexPair
        If _matches.Count = 0 Then Return MatchIndexPair.Empty

        For i As Integer = 0 To _matches.Count - 1
            Dim match = _matches(i)
            If Not matchday.ContainsPlayerIn(match) Then
                Return match
            End If
        Next

        Return MatchIndexPair.Empty
    End Function
End Class

File: POFF.Kicker/Domain/MatchGenerators/MatchdaysMatchGenerator.cs
using System.Collections.Generic;
using System.Linq;

namespace POFF.Kicker.Domain.MatchGenerators;


public class MatchdaysMatchGenerator : IMatchGenerator
{
    private readonly int _teamsCount;
    private readonly List<MatchIndexPair> _matches = [];
    private readonly List<Matchday> _matchdays = [];

    public MatchdaysMatchGenerator(int teamsCount = 10)
    {
        _teamsCount = teamsCount;
    }

    public IEnumerable<MatchIndexPair> Generate()
    {
        _matches.Clear();
        _matches.AddRange(GenerateMatches());
        _matchdays.Clear();
        _matchdays.AddRange(GenerateMatchdays());
        return MatchIndexPairsOrderdByMatchdays();
    }

    private IEnumerable<MatchIndexPair> MatchIndexPairsOrderdByMatchdays()
    {
        foreach (var matchday in _matchdays)
        {
            foreach (var matchIndexPair in matchday)
                yield return matchIndexPair;
        }
    }

    private IEnumerable<MatchIndexPair> GenerateMatches()
    {
        for (int i = 1, loopTo = _teamsCount; i <= loopTo; i++)
        {
            for (int j = i + 1, loopTo1 = _teamsCount; j <= loopTo1; j++)
                yield return new MatchIndexPair(i, j);
        }
    }

    private IEnumerable<Matchday> GenerateMatchdays()
    {
        int blockSize = _teamsCount / 2;

        for (int i = 0, loopTo = _teamsCount - 2; i <= loopTo; i++)
            yield return GenerateMatchday(blockSize);
    }

    private Matchday GenerateMatchday(int blockSize)
    {
        var matchday = new Matchday();
        while (matchday.Count < blockSize)
        {
            var invalidatedMatches = new List<MatchIndexPair>();
            var nextMatch = GetNextMatch(matchday);
            if (nextMatch.Equals(MatchIndexPair.Empty))
            {
                nextMatch = _matches.Except(invalidatedMatches).FirstOrDefault();
                if (nextMatch is null)
                    nextMatch = MatchIndexPair.Empty;
                invalidatedMatches.AddRange(matchday.MatchesWithPlayersFrom(nextMatch));
                _matches.AddRange(invalidatedMatches);
                matchday.RemoveAll(m => invalidatedMatches.Contains(m));
            }
            matchday.Add(nextMatch);
            _matches.Remove(nextMatch);
        }
        return matchday;
    }

    private MatchIndexPair GetNextMatch(Matchday matchday)
    {
        if (_matches.Count == 0)
            return MatchIndexPair.Empty;

        for (int i = 0, loopTo = _matches.Count - 1; i <= loopTo; i++)
        {
            var match = _matches[i];
            if (!matchday.ContainsPlayerIn(match))
            {
                return match;
            }
        }

        return MatchIndexPair.Empty;
    }
}